### PR TITLE
WAR added to fuse native_dropout in DifferentiableGraph

### PR DESCRIPTION
**Pull request by jjsjann123**
1. inserted a quick pass to populate tensor type from native_dropout output tensor to output mask;
2. allow fusion of producer with multiple outputs in graph partitioner;

### CODE REVIEW

**Comment by jjsjann123** [comment]
dev_branch is broken. Let's not merge anything there yet.
I somewhat tested this guy on a previous dev_branch. Opening this PR just so @kevinstephano could move forward with bert demo.

**Comment by jjsjann123** [comment]
Forgot to add a comment here that this PR should be good for review and merge.

File: test/test_jit_cuda_fuser.py [hunk 1]
```diff
@@ -1750,6 +1750,18 @@ def t(x: torch.Tensor, p: float, train: bool):
         # numbers between eager mode and the jit is different
         self._run_training_helper(t_jit, t, grads, x, 0.0, True)
 
+        def t2(x: torch.Tensor, p: float, train: bool):
+            o = torch.nn.functional.softmax(x, dim=-1)
+            o = torch.nn.functional.dropout(o, p, training=train)
+            return o
+
+        t2_jit = torch.jit.script(t2)
```
Comment (tlemo) [inline test/test_jit_cuda_fuser.py:1758]: just for my own education: what is the reason for not using @torch.jit.script decorator directly on t2?

File: torch/csrc/jit/codegen/cuda/partition.cpp [hunk 2]
```diff
@@ -148,26 +148,21 @@ bool isNonBroadcastElementWise(const Node* n) {
...
-        if (maybeBroadcast(t_type, n_output_shape)) {
-          return false;
+  for (const auto output : n->outputs()) {
+    auto n_output_type = output->type()->cast<TensorType>();
+
+    // TODO: we need to stay on safer side instead of "default to return true
+    // when shape information is not available.", Change that when we enable
+    // profiling on autodiff FW execution.
+    if (n_output_type != nullptr && n_output_type->sizes().sizes()) {
+      std::vector<c10::optional<int64_t>> n_output_shape =
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/partition.cpp]: nit : can we use const vector<...>& here?

File: torch/csrc/jit/codegen/cuda/graph_fuser.cpp [hunk 3]
```diff
@@ -1358,6 +1358,23 @@ void decomposeLinearOps(Block* block) {
...
+// This is temporary to handle intermediate tensor inserted by autodiff is not
+// being profiled
+void markMissingType(Block* block) {
+  std::vector<Node*> linear_nodes;
+  for (Node* n : block->nodes()) {
+    for (Block* b : n->blocks()) {
+      markMissingType(b);
+    }
+    // fill in the tensor type for mask output in `aten::native_dropout`
+    if (n->kind() == aten::native_dropout) {
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/graph_fuser.cpp]: this seems a bit fragile: we're just checking the kind w/o considering the schema - in the unlikely case we get a different another native_dropout overload or the schema changes the hardcoded outputs()[1] and [0] may not be what we expect.

am I overthinking it?

**Review by tlemo — APPROVED** [review]

File: torch/csrc/jit/codegen/cuda/graph_fuser.cpp (hunk 3 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/graph_fuser.cpp]: Good catch and you are not over thinking this. Let me switch to a schema match just to be on the safer side.

File: test/test_jit_cuda_fuser.py (hunk 1 above)
Comment (jjsjann123) [reply test/test_jit_cuda_fuser.py:1758]: there's no good reason for that. I just never default to decorator (not a sophisticated pythoner)

**Comment by jjsjann123** [comment]
Passed python tests after ToT devel branch merge. I'm merging this PR after CI.
